### PR TITLE
Graph.add_node, Graph.add_edge, and Graph.add_edges_from modification

### PR DESCRIPTION
**Pull request by thegreathippo**
For each respective method, I just snuck in an empty `node_attr` and `edge_attr` dictionary, updated it based on the supplied parameters (`attr_dict` and `attr`), then either plugged it in if no attribute dictionary existed -- or updated an existing attribute dictionary if it was there. That way, you avoid mutating attr_dict and don't end up accidentally sharing a dictionary across multiple nodes (or other objects). 

Notably, this caused some test failures -- because we're no longer invoking the `update` method on` attr_dict`. And since a Python dictionary's `update` method can accept an empty sequence, just passing an empty sequence through `attr_dict` no longer raised an exception. So I added an int to the sequences to make the exception raise.

This leads to an interesting positive trade-off, though: Since we're updating a dictionary with `attr_dict`, `attr_dict` doesn't _have_ to be a dictionary. It can be any value that the `update` method accepts: Including sequences of 2-d tuples. So this:

```
>>> import networkx
>>> g = networkx.Graph()
>>> g.add_node("node", [("weight", 5)])
>>> print(g.node["node"])
{'weight' : 5}
```

-- is perfectly acceptable. So is:

```
>>> import networkx
>>> g = networkx.Graph()
>>> my_dict = {"x": 0}
>>> g.add_node("node", my_dict.items())
>>> print(g.node["node"])
{'x' : 0}
```

And, of course, the old way works too, sans mutation or direct assignment:

```
>>> import networkx
>>> g = networkx.Graph()
>>> my_dict = {"x": 0}
>>> g.add_node("node", my_dict, x=1)
>>> print(g.node["node"])
{'x' : 1}
>>> print(my_dict)
{'x' : 0}
```

(As an aside, this is my first attempt at trying to help with a collaborative github project -- so if I've messed up anywhere in this process, please feel free to tell me!)

### CODE REVIEW

**Comment by hagberg** [comment]
Thanks for the pull request.  This code is part of the core of networkx so it may take us a little while to figure out how we want to proceed (and why).   I'd especially like comments from @dschult.  What is the utility of having attr_dict unless we actually want side effects like we are trying to prevent here?

**Comment by dschult** [comment]
I can see that this needs more explicit documentation... 

Here is what I recall from our original thinking:
Sometimes we want attr_dict to actually BE the object used to store attributes. If you have a group of nodes that should all have the same attributes, it might be useful to have a single object to update all those attributes. Also, if you have multiple graphs using the same node, it would be useful to be able to change attributes for a node in one graph and have it change that node's attributes in all the graphs. 

Other times we only want the information held in attr_dict to be stored. You might set all nodes to have color "blue" initially, but still want to allow the color to change for each node separately later. Then we don't want to use the same object to set the initial colors. We could use a single dict to set the initial attribute values without having them all use the same object to store those attributes.

We also had to decide the relative order of updates for attr_dict and the keyword dict called attr. And they aren't the same for ```add_nodes_from``` and ```add_node```.

I should also remark that we paid much more attention to this for edges than for nodes. It is possible that node attributes have been relatively neglected in this regard.

**Comment by dschult** [comment]
**Summary of Existing Code**
It looks like for new nodes:

    add_node(n, attr_dict, **attr)   use the object attr_dict updated by attr
    add_nodes_from( [(n,adict)], **attr)  use a copy of attr updated by adict

For updating existing nodes, the existing attribute dictionary is updated as:

    add_node(n, attr_dict, **attr)   update by attr_dict then by attr
    add_nodes_from( [(n,adict)], **attr)    update by attr then by adict

For edges, there is the option of a custom ```edge_attr_dict_factory```, but let's assume that is ```dict```.
For new edges:

    add_edge(u, v, adict, **attr)  new dict updated by adict then by attr
    add_edges_from( [(u, v, dd)], adict, **attr)  new dict updated by adict then by attr then by dd

For existing edges:

    add_edge(u, v, adict, **attr)  update by adict then by attr 
    add_edges_from( [(u, v, dd)], adict, **attr)  update by adict then by attr then by dd

**Comment by dschult** [comment]
I think this PR is a good one. It reduces surprise mutations of input ```attr_dict``` when calling ```add_node``` and/or ```add_edge```.

The downside is that users of ```add_node``` cannot reuse the same attribute dict for multiple nodes or for the same node in multiple graphs.  But I don't think that is a common usage. It is also still possible to do so via direct assignment to ```G.node```. 

I'll go through the PR and look for other possible issues/trouble.  Thanks for bringing this up!

File: networkx/classes/graph.py [hunk 1]
```diff
@@ -852,14 +854,15 @@ def add_edges_from(self, ebunch, attr_dict=None, **attr):
...
+        edge_attr = {}
         if attr_dict is None:
-            attr_dict = attr
-        else:
-            try:
-                attr_dict.update(attr)
-            except AttributeError:
+            attr_dict = {}
+        try:
+                edge_attr.update(attr_dict)
```
Comment (jfinkels) [inline networkx/classes/graph.py]: Wrong indentation level, should be four spaces instead of eight.

Comment (jfinkels) [inline networkx/classes/graph.py]: This applies below and above as well.

File: networkx/classes/graph.py [hunk 2]
```diff
@@ -852,14 +854,15 @@ def add_edges_from(self, ebunch, attr_dict=None, **attr):
         >>> G.add_edges_from([(3,4),(1,4)], label='WN2898')
         """
         # set up attribute dict
+        edge_attr = {}
         if attr_dict is None:
```
Comment (dschult) [inline networkx/classes/graph.py]: Can we do a similar change to this structure as with my comment above? Use ```if attr_dict is not None:``` and proceed from there?

File: networkx/classes/graph.py (hunk 1 above)
Comment (dschult) [inline networkx/classes/graph.py]: Let's change this to:

    if attr_dict is not None:
        try:
            edge_attr.update(attr_dict)
        except TypeError:
            msg = "The attr_dict argument must be a dict or iterable of 2-tuples"
            raise NetworkXError(msg)

This removes the whole try/except portion when ```attr_dict is None``` (which happens a lot).

**Comment by hagberg** [comment]
How about eliminating the attr_dict optional keyword completely?  That would make the methods simpler.  

Also #1583 is still open and maybe relevant if we consider change the function arguments.

**Comment by dschult** [comment]
+1 for eliminating ```attr_dict``` argument.  

Users can use ```G.add_node(1, color="b", **attr_dict)```  without trouble so long as keys don't collide.

**Comment by thegreathippo** [comment]
Also, if you _really_ need to have nodes share attributes (or have nodes rely on a dictionary outside the graph for their attributes), you can always directly assign a dictionary to a node:

```
>>> import networkx
>>> g = networkx.Graph()
>>> g.add_node("node")
>>> my_dict = {"x" : 0}
>>> g.node["node"] = my_dict
>>> my_dict["x"] = 1
>>> print(g.node["node"])
{'x' : 1}
```

If it's okay, I'll see if I can rewrite the method to just use kwargs, rather than attr_dict? Keeping in mind, at a glance, it means I'll have to update a few error tests? (If you'd rather handle it yourselves, I'd understand; I'm a bit new at this!)

(I also think it'd be easy to tie this into a solution that gets rid of the u and v positional terms in a way that's fairly elegant, but I want to make sure I can actually get rid of attr_dict without breaking anything, first)

**Comment by dschult** [comment]
I'm afraid that removing the argument ```attr_dict``` will introduce errors all over the codebase. You can try, but I'm pretty sure there will be many places where you have to add ```**```. I also think there are a few places where we use an attr_dict that has numbers for keys.  With **attr the dict has to have strings as keys.
  
I'd be interested to hear about attempts to remove the u and v positional terms.

**Comment by thegreathippo** [comment]
I see what you mean -- especially when you consider that a direct assignment to the `attr_dict` keyword parameter is actually totally valid even if you remove `attr_dict` as a positional; now, you're adding the string 'attr_dict' as a _node attribute_, which creates all sorts of weird errors. I'm still interested in seeing if I can remove / simplify all the instances of that problem, though I understand it may be well above my proficiency level!

(Also, as another aside -- since the impression I got was that node and edge attributes are meant to simulate Python attributes, removing the ability to directly pass a dictionary to `add_node`, `add_edge`, and `add_edges_from` prevents you from assigning non-strings to these things -- which better simulates Python attribute assignment)

Removing n is easy if you get rid of `attr_dict`; just make it `*args` and add every value in `args` as a node; maybe raise a `TypeError` `if len(args) == 0`. Removing u-v seems similarly easy (I think?): Make them `*args`, raise a `TypeError` `if len(args) < 2`, and add an edge with the given attributes between every node listed in `*args`.

So,` mygraph.add_edge("node1", "node2", "node3")` would create 3 edges: node1->node2, node2->node3, node3->node1.

(Of course, this is more `add_edges` than `add_edge`, but if you really wanted to keep `add_edge` singular, just write `add_edges`, and wrap it in `add_edge` -- which raises an exception `if len(args) != 2`)

(Also, for the record, not sure what sort of consequence this would have on directgraphs or multigraphs)

**Comment by hagberg** [comment]
That doesn't seem so bad.  The situation with multigraphs and directed graphs should be very similar. 
It might be a little harder with `add_edges` and `add_edges_from` since my feeling is the attr_dict pattern is used a lot more in those cases.  I might be wrong.

About the function naming (plural/singular) and *args approach.  I agree that there may be some opportunities to make the interface simpler.  It might be a little to out-of-scope for this release but maybe there is a clean way to handle backwards compatibility if we decide we want to proceed.

**Comment by thegreathippo** [comment]
Sorry about the pre-emptive commit earlier; still getting a handle on how pull requests work -- this last commit updates `add_node`, `add_edge`, and `add_edges_from` for the `Graph`, `MultiGraph`, `MultiDiGraph`, `DiGraph`, and `timingclasses.py` with the new setup suggested above. I also hunted down all the functions that passed dictionaries to these methods and made them either assign keywords directly or indirectly via ** (this way, if there ever comes a time to get rid of attr_dict, it should be much easier to do). One exception in [`convert.py`](https://github.com/networkx/networkx/blob/master/networkx/convert.py):

```
    else: # not a multigraph to multigraph transfer
        if G.is_multigraph() and not G.is_directed():
            # d can have both representations u-v, v-u in dict.  Only add one.
            # We don't need this check for digraphs since we add both directions,
            # or for Graph() since it is done implicitly (parallel edges not allowed)
            seen=set()
            for u,nbrs in d.items():
                for v,data in nbrs.items():
                    if (u,v) not in seen:
                        G.add_edge(u,v,attr_dict=data)
                    seen.add((v,u))
        else:
            G.add_edges_from( ( (u,v,data)
                                for u,nbrs in d.items()
                                for v,data in nbrs.items()) )
    return G
```
As @dschult mentioned above, this is a case where you're passing a dictionary with non-string keys -- so it can't be passed as keyword arguments. I tried just assigning the edge and updating the dictionary after-the-fact, but this raised all sorts of weird errors; a few other attempts didn't work, either. I think this is probably because the graph in question is a multigraph -- and I have no idea how to properly access their edge-attribute dictionaries.

I'm also a little concerned about this change, in [`convert_matrix.py`](https://github.com/networkx/networkx/blob/master/networkx/convert_matrix.py):

```
        for row in df.values:
            g.add_edge(row[src_i], row[tar_i], **{i:row[j] for i, j in edge_i})
```
--because while it passes all the tests, I'm not really sure what this is doing or what it's for -- and I'm concerned that in the wild, it might be expected for this call to be able to handle numbers as edge-attributes.

Other than that, I think I got them all? The PR is a bit of a mess on account of me deleting a bunch of code experimentally, then committing -- I re-added it, but now it's flagged as changed. If that makes it a problem to review, I can close this pull request, start a new one, and copy over my edited files so the things I changed back don't get flagged as having been changed.

**Comment by jfinkels** [comment]
@thegreathippo If you wish to edit (or "squash") the commits you've made on your branch, you can use `git rebase`; see "[About Git rebase][1]" for more information.

[1]: https://help.github.com/articles/about-git-rebase/

**Comment by dschult** [comment]
Nice work -- I think you are right that the line from ```convert_matrix.py``` would potentially cause trouble when integers show up as an edge attribute key.  Perhaps that one should add the edge and then update:

    g.add_edge(row[src_i], row[tar_i])
    g[row[src_i]][row[tar_i]].update((i, row[j]) for i, j in edge_i)

The function declaration and doc_string for the function ```add_edges_from``` in ```graph.py``` seems to be duplicated.  Also It doesn't look like you have removed the argument ```attr_dict```, though you have laid the groundwork for doing so.

I'll look more later.

**Comment by hagberg** [comment]
This looks like a good start.  Let's first focus on getting the `add_node`, `add_nodes_from`, `add_edge`, and `add_edges_from` functions the way we want them (completely removing the attr_dict argument).

**Comment by thegreathippo** [comment]
Pardon; I mistakenly got the impression that we might not want to remove the `attr_dict` argument _completely_ (on account of breaking backward compatibility; someone who called these methods and assigned a dictionary via `attr_dict` would end up assigning that dictionary to a note attribute _called_ `attr_dict`, which would potentially create really-hard-to-track errors -- or worse still, would corrupt data without raising any exceptions at all!). But with this last commit, it should be pretty trivial to remove the `attr_dict` completely -- I'll work on it when I get home.

(Though I can't _actually_ remove it until I find a way to fix the `attr_dict` assignment in `convert.py`; any thoughts on how I can safely get rid of that? Otherwise, I'll do a little reading on how multigraph edge attributes work to see if I can figure it out)

I'll also modify `convert_matrix.py` to use the update method, instead -- as for the duplicate function and documentation, I'll take a look at that, too.

**Comment by dschult** [comment]
I think the best way to change ```convert.py``` is to use:

    G.add_edge(u,v)
    G.edge[u][v].update(data)  # use update instead of add_edge to allow non-string keys

**Comment by hagberg** [comment]
I was looking at convert.py a little.  It breaks the tests in a place where a slightly more ugly fix is needed - e.g.

```python
            for u,nbrs in d.items():
                for v,data in nbrs.items():
                    if (u,v) not in seen:
#                        G.add_edge(u,v,**data) # can't do this with integer keys
                        G.add_edge(u,v,key=0)
                        G[u][v][0].update((k,v) for k,v in data.items())
                    seen.add((v,u))
        else:
```

**Comment by thegreathippo** [comment]
Thanks; I think that'll work perfectly! I now get why my attempts weren't working before -- I didn't realize until recently that multigraph edges use keys (makes sense, in retrospect).

Also, `convert.py` has a few more assignment-via-dict cases that I missed:
```
    if multigraph_input:
        # make a copy of the list of edge data (but not the edge data)
        if G.is_directed():
            if G.is_multigraph():
                G.add_edges_from( (u,v,key,data)
                                  for u,nbrs in d.items()
                                  for v,datadict in nbrs.items()
                                  for key,data in datadict.items()
                                )
            else:
                G.add_edges_from( (u,v,data)
                                  for u,nbrs in d.items()
                                  for v,datadict in nbrs.items()
                                  for key,data in datadict.items()
                                )
        else: # Undirected
            if G.is_multigraph():
                seen=set()   # don't add both directions of undirected graph
                for u,nbrs in d.items():
                    for v,datadict in nbrs.items():
                        if (u,v) not in seen:
                            G.add_edges_from( (u,v,key,data)
                                               for key,data in datadict.items()
                                              )
                            seen.add((v,u))
            else:
                seen=set()   # don't add both directions of undirected graph
                for u,nbrs in d.items():
                    for v,datadict in nbrs.items():
                        if (u,v) not in seen:
                            G.add_edges_from( (u,v,data)
                                        for key,data in datadict.items() )
                            seen.add((v,u))
```
--which I strongly suspect should all allow for non-string attribute assignment, too. I should be able to fix the multigraph cases using a similar solution as above. I'll also do one more (slower) comb through the code to make sure I didn't miss anything else.

(Also, realizing only now that I should keep my fork sync'd with the master; I'll work on setting that up when I get home, too)

(Also, thank you all for helping walk me through this! I'm learning a lot, and I'm thrilled to have a chance to contribute something to this project!)

**Comment by hagberg** [comment]
Note that my code above is a hack forcing 0 as the key for every multigraph edge.  Since the for that section input isn't a multigraph this would be ok since there are not supposed to be duplicate edges in the output.  The convert.py code is a bit tricky to handle the multiple graph classes we are currently using.

**Comment by thegreathippo** [comment]
Yeah, I was looking at that, and the challenge multigraph edges produce in general -- it strikes me that having `add_edge` actually _return_ the edge key (for the case of multigraphs) would solve a lot of those problems. Or, at least, return the edge attribute dictionary.

(Though, if you did that, it would be reasonable to expect every `add_edge` method to return something similar, and then every `add_node` method, and `add_edges_from`, and now I'm getting into a whole other deal that's way outside the scope of this fix)

**Comment by dschult** [comment]
Another way to alleviate some of the multigraph pain would be to require an explicit edge key for each edge.  I think there was a ticket suggesting that ```add_edge``` return the edge key some time ago. Not sue whether it is still active -- it was part of another discussion.

**Comment by thegreathippo** [comment]
Ticket #1376, it looks like? My concern re: returning the edge key was that you currently don't return anything for _any_ of the methods that add nodes and edges -- and there's an elegant utility to having them all return the same thing (the attribute dict for the node/edge -- or a dictionary-of-attribute-dicts in the case of `add_nodes_from` and `add_edges_from`). It certainly makes the fix for `convert.py` much prettier (`G.add_edge(u,v).update((k, v) for k, v in data.items())`). But I can see how returning the edge key makes dealing with multigraphs a lot simpler; either way, it's probably way outside the scope of this fix -- but I still wanted to mention it.

Also, I didn't get a chance to touch the code last night (got tied up); I should be able to finish this up over the weekend.

**Comment by hagberg** [comment]
The "return edge key" issue is indeed unresolved and a little tricky to change.  If we can proceed here without opening that up that would be good.

**Comment by thegreathippo** [comment]
I removed all references to attr_dict, but I keep getting this error on all the travis builds that use miniconda:
```
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python3.3/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.3/site-packages/networkx/tests/test_convert_pandas.py", line 44, in test_from_dataframe_all_attr
    self.assert_equal(MG, MGtrue)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.3/site-packages/networkx/tests/test_convert_pandas.py", line 32, in assert_equal
    assert_true( nx.is_isomorphic(G1, G2, edge_match=lambda x, y: x == y ))
nose.proxy.AssertionError: False is not true
    'False is not true' = self._formatMessage('False is not true', "%s is not true" % safe_repr(False))
>>  raise self.failureException('False is not true')
```
In tracking this error down, I figured out it's rooted in `convert_matrix.py`, [here](https://github.com/thegreathippo/networkx/blob/a1af38a7205e54abc5ea60c919ba009f2e1dcccc/networkx/convert_matrix.py#L212-L218):
```
        for row in df.values:
            if g.is_multigraph():
                g.add_edge(row[src_i], row[tar_i], key=0)
                g[row[src_i]][row[tar_i]][0].update((i, row[j]) for i, j in edge_i)                
            else:
                g.add_edge(row[src_i], row[tar_i])
                g[row[src_i]][row[tar_i]].update((i, row[j]) for i, j in edge_i)
```
When I allow a multigraph's `add_edge` method to return the attribute dictionary of the edge we're adding (to bypass having to acquire the appropriate key), then update those attributes here, the error disappears -- but is replaced with all sorts of errors that emerge because several sample cases expect the `add_edge` method to return nothing). So I'm pretty certain this is the culprit.

I'm not sure how else to fix this, though -- any thoughts would be appreciated. Other than this, everything seems to work just fine; if I can get this last error to clear, it looks like it should pass through.

**Comment by hagberg** [comment]
I don't think we should return anything from `add_edge`.  I know that means some engineering work-around for this case but that fix is better to postpone since it opens up a lot of other design tweaks.

**Comment by dschult** [comment]
The problem stems from using ```key=0``` for all edges. If the dataframe has multiple rows with the same source and target we don't get multiedges this way.

Luckily ```from_pandas_dataframe``` doesn't have a way to specify keys, so we KNOW the default is being used. That means the most recent edge key is the maximum of all edge keys for that pair of nodes.  We can figure out the ```key``` without having it returned by ```add_edge```.  Try this:

       for row in df.values:
            s, t = row[src_i], row[tar_i]
            if g.is_multigraph():
                g.add_edge(s, t)
                key = max(g[s][t])  # default keys just count, so max is most recent 
                g[s][t][key].update((i, row[j]) for i, j in edge_i)
            else:
                g.add_edge(s, t)
                g[s][t].update((i, row[j]) for i, j in edge_i)

**Comment by thegreathippo** [comment]
Thanks; I appreciate it -- that did it! It's finally passing through the tests. I think that's everything, then?

**Comment by hagberg** [comment]
Ok, thanks!  I added the "needs review" tag and we should take a close look at the changes since this touches 31 files including the core graph classes.

**Comment by hagberg** [comment]
I looked through all of the changes and they seem fine to me.  

The multigraph key implementation we are using definitely needs more thought and redesign.  The current code is OK but that will come back to bite us some other day...

**Comment by thegreathippo** [comment]
Oh, crap -- I just looked it over myself and realized I forgot to take `attr_dict` out of `timingclasses.py` (it still uses my first solution, which was to copy it over to a whole new dictionary). I'll take care of that real quick when I get home, it should be a fast and easy fix.

**Comment by hagberg** [comment]
There are two other things I thought of 
1) We could make a note about this in https://github.com/networkx/networkx/blob/master/doc/source/reference/release_2.0.rst
2) I didn't look closely but it should be documented in the `*_from()` methods that if an attribute is specified with keywords it is applied before any node-specific attribute in the container of nodes or edges.

File: networkx/readwrite/json_graph/adjacency.py [hunk 3]
```diff
@@ -145,15 +145,18 @@ def adjacency_graph(data, directed=False, multigraph=True, attrs=_attrs):
...
             target = target_data.pop(id_)
             if not multigraph:
-                graph.add_edge(source, target, attr_dict=tdata)
+                graph.add_edge(source, target)
+                graph[source][target].update(tdata)
             else:
                 ky = target_data.pop(key, None)
-                graph.add_edge(source, target, key=ky, attr_dict=tdata)
+                graph.add_edge(source, target, key=ky)
+                graph[source][target][ky].update(target_data)
```
Comment (dschult) [inline networkx/readwrite/json_graph/adjacency.py]: This ```target_data``` should be ```tdata``` I believe.  

Also, in this function, we copy the ```tdata``` dict just to ```pop(id_)``` from it without affecting the original dict.  Seems like instead we should use ```target = tdata[id_]```.  But that change should apply to both the node and edge portions of the code and is better done in another PR.

**Comment by dschult** [comment]
As for timingclasses.py, that file was created by me to ease testing the impact on speed of release v2.  It probably shouldn't even be there, but is a convenience for me (stored on github).  We shouldn't make changes to the API in that file or we won't be able to compare the speed of the two APIs.

Could you please take out the changes you make to timingclasses.py?  Sorry about the confusion.

**Comment by dschult** [comment]
I'll open another issue to discuss multigraph ```key```
I think @hagberg is correct that it will come back to bite us someday.

**Comment by thegreathippo** [comment]
Sorry about the huge delay on getting back to this -- I've made the requested changes, but now I'm getting an error in a place I completely didn't expect:
```

======================================================================
FAIL: test_match_helpers.TestGenericMultiEdgeMatch.test_generic_multiedge_match
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/networkx/algorithms/isomorphism/tests/test_match_helpers.py", line 46, in test_generic_multiedge_match
    assert_false(id_match(self.G3[3][4], self.G4[4][5]))
AssertionError: True is not false
    'True is not false' = self._formatMessage('True is not false', "%s is not false" % safe_repr(True))
>>  raise self.failureException('True is not false')
```
All I can figure is that somehow, the timingclasses.py is causing this (since the error only pops up when I first revert the timingclasses.py to its previous form). I'm not sure how to handle this, because I'm completely unfamiliar with nose's testing apparatus. 

Also, I'm leery about editing the release document myself, mostly because I'm no technical writer and I'm not familiar with the etiquette (should I write the notation at the top? At the bottom? Somewhere in the middle?). I can give it a shot if you'd like, but it's going to be a shot in the dark.

**Comment by dschult** [comment]
The Travis CI error you are seeing is occurring because the maxflow problem seems to have two solutions and the algorithm is now finding the one that isn't checked for. I believe this is a problem with the test not accounting for the arbitrary ordering of dict.  This test was touched recently (in May) and we might not have taken into account this potential test failure.  I'm not sure why it would show up on the commit where you changed timingclasses.py because it doesn't use that code at all. Perhaps something else (like a different version of some package on Travis) changed the order of dict reporting for this particular case.  Anyway, don't worry about that for this PR.

I'll put together something for the release.  I'm also going to take one last pass through the PR before merging.

**Comment by dschult** [comment]
Ahhh.... I spoke too soon (as I often do :)

In test_match_helpers.py the ```setup``` method uses ```attr_dict=attr_dict``` instead of ```**attr_dict```.  So we were ending up with an attribute called ```attr_dict```.  Still not sure why that passed previous tests.

I also notice that algorithms/tree/branchings.py creates a class with methods ```add_edge``` and ```add_edges_from```. Those methods should remove the ```attr_dict``` keyword to match the main classes.

**Comment by dschult** [comment]
In classes/multidigraph.py the method ```add_edge``` should remove the parameter ```attr_dict``` from the doc_string.

**Comment by chebee7i** [comment]
The change in `branchings.py` should be ok to do. None of the code actually uses `attr_dict`.

**Comment by dschult** [comment]
I'll add to this PR and merge shortly.
